sqlmesh/core/context.py

```diff
@@ -372,14 +378,39 @@ def __init__(
         self.concurrent_tasks = concurrent_tasks or self._connection_config.concurrent_tasks
         self._engine_adapter = engine_adapter or self._connection_config.create_engine_adapter()
 
+        self._snapshot_evaluators: UniqueKeyDict[str, SnapshotEvaluator] = UniqueKeyDict(
+            "evaluators"
+        )
+        self._default_evaluator: t.Optional[SnapshotEvaluator] = None
+        self._test_connection_configs: UniqueKeyDict[str, ConnectionConfig] = UniqueKeyDict(
+            "test_configs"
+        )
+
+        if not gateway:
+            for gateway_name in self.config.gateways:
+                connection = self.config.get_connection(gateway_name)
+                adapter = (
+                    connection.create_engine_adapter()
+                    if gateway_name != self.config.default_gateway
+                    else self._engine_adapter
+                )
+                evaluator = SnapshotEvaluator(
```

Comment:
Hmm, I think we'd still want to manage all the snapshots within a single evaluator but direct them to the relevant engine_adapter's as required.

I can see why youve done this though. The problem here though is that the `engine_adapter` is an instance property of the `SnapshotEvaluator`, so the `SnapshotEvaluator` currently assumes a single gateway / engine.

However, I dont see anything engine specific within `SnapshotEvaluator`  so perhaps it could take a function that returns an EngineAdapter based on the supplied gateway name and use that to acquire the correct engine adapter when it needs to materialize a snapshot.

@izeigerman i'd be keen to hear your thoughts on this

Reply:
i think it could be cleaner if there was only one snapshot evaluator with many adapters, because the rest of the code base only deals with one evaluator

Reply:
right i will change then to use multiple engine adapters in a single evaluator. I went back and forth on the best approach, especially considering parts involving concurrency or areas that would require engine specific logic such as during schema creation that required more precise changes in the evaluator. but overall it does seem that using multiple engines would be the cleaner solution, so currently refactoring the code accordingly